### ISSUE TITLE
FaceRecognition integration doesn't work with postgresql database

### ISSUE THREAD

**Issue by quadragan** [pulsejet/memories#304]
Hello,

There seems to be an issue with the integration of the Face Recognition app when running on postgresql database
This appears in log when trying to open "People (Face Recognition)" tab
Works okay with MariaDB

OC\DB\Exceptions\DbalException: An exception occurred while executing a query: SQLSTATE[42803]: Grouping error: 7 ERROR: column "frp.id" must appear in the GROUP BY clause or be used in an aggregate function LINE 28: ) SELECT "frp"."id", "frp"."user" as "user_id", "frp"."n... ^
/var/www/nextcloud/lib/private/DB/ConnectionAdapter.php - line 69:

OC\DB\Exceptions\DbalException::wrap()

/var/www/nextcloud/apps/memories/lib/Db/TimelineQueryDays.php - line 328:

OC\DB\ConnectionAdapter->executeQuery()

/var/www/nextcloud/apps/memories/lib/Db/TimelineQueryPeopleFaceRecognition.php - line 111:

OCA\Memories\Db\TimelineQuery->executeQueryWithCTEs()

/var/www/nextcloud/apps/memories/lib/Controller/PeopleController.php - line 133:

OCA\Memories\Db\TimelineQuery->getPeopleFaceRecognition()

/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php - line 225:

OCA\Memories\Controller\PeopleController->facerecognitionPeople()

/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php - line 133:

OC\AppFramework\Http\Dispatcher->executeController()

/var/www/nextcloud/lib/private/AppFramework/App.php - line 172:

OC\AppFramework\Http\Dispatcher->dispatch()

/var/www/nextcloud/lib/private/Route/Router.php - line 298:

OC\AppFramework\App::main()

/var/www/nextcloud/lib/base.php - line 1047:

OC\Route\Router->match()

/var/www/nextcloud/index.php - line 36:

OC::handleRequest()

**Comment by matiasdelellis** [comment]
Hi @quadragan 
Please, test last update of Face Recognition app.

We have reports that a migration in the previous version don't work in PostgreSQL.

**Comment by quadragan** [comment]
Hi,
Still getting the same errors after the update. Or is there some action that should be done first?

**Comment by pulsejet** [comment]
@matiasdelellis postgres requires all selected fields in a group by, but we only have either `frp.name` or `frp.id` in this case .. maybe be need two different queries, not sure which fields we need and which are redundant.

**Comment by zealot128** [comment]
I havn't used PHP for years, but I am quite familiar with PG.
I've tried to debug the sql statement a little.

<details>

<summary>SQL statement generated by the TimelineQuery</summary>

```sql
WITH RECURSIVE oc_cte_folders_all(fileid, rootid) AS (
                SELECT
                    f.fileid,
                    f.fileid AS rootid
                FROM
                    oc_filecache f
                UNION ALL
                SELECT
                    f.fileid,
                    c.rootid
                FROM
                    oc_filecache f
                INNER JOIN oc_cte_folders_all c
                    ON (f.parent = c.fileid
                        AND f.mimetype = (SELECT "id" FROM "oc_mimetypes" WHERE "mimetype" = 'httpd/unix-directory')
                        AND f.name <> '.archive'
                    )
            ), oc_cte_folders AS (
                SELECT
                    fileid,
                    MIN(rootid) AS rootid
                FROM
                    oc_cte_folders_all
                GROUP BY
                    fileid
            )
        SELECT "frp"."id", "frp"."user" as "user_id", "frp"."name", COUNT(DISTINCT m.fileid) AS "count"
        FROM "oc_facerecog_persons" "frp"
        INNER JOIN "oc_facerecog_faces" "frf" ON "frp"."id" = "frf"."person"
        INNER JOIN "oc_facerecog_images" "fri" ON "fri"."id" = "frf"."image"
        INNER JOIN "oc_memories" "m" ON ("fri"."file" = "m"."fileid")
        INNER JOIN "oc_filecache" "f" ON "f"."fileid" = "m"."fileid"
        INNER JOIN "oc_cte_folders" "cte_f" ON "f"."parent" = "cte_f"."fileid"
        WHERE ("frp"."name" IS NOT NULL) AND ("frp"."is_visible" = 't')
        GROUP BY "frp"."name" 
        ORDER BY "count" DESC, "name" ASC, "frp"."id" ASC
```

(I've removed the Bind parameters $1..$4)

</details>

Seems to be, it can just be Grouped by id anyways (frp.id and name are in the same table, grouping by id should be the same, PG can then figure out, that you want the name of the same row, and it will work!)

```
GROUP BY "frp"."id"
```

So switching: ``lib/Db/TimelineQueryPeopleFaceRecognition.php``

```php

        if ($show_clusters) {
            // GROUP by ID of face cluster
            $query->groupBy('frp.id');
            $query->where($query->expr()->isNull('frp.name'));
        } else {
            // GROUP by name of face clusters
            // $query->groupBy('frp.id');  <---- doesnt work
            $query->groupBy('frp.id');  // <---------------------
            $query->where($query->expr()->isNotNull('frp.name'));
        }
```

---

THEN i get a new error! Yeah Progress! :) 

```
OC\DB\Exceptions\DbalException: An exception occurred while executing a query: 
SQLSTATE[42703]: Undefined column: 7 
ERROR: column "count" does not exist LINE 28: ...rp"."is_visible" = $3) GROUP BY "frp"."id" HAVING "count" > ... ^
```

Seems to be this, that generates the invalid SQL:

```php
$query->having($query->expr()->gt('count', $query->createNamedParameter(1)));
```

which produces -> 

```sql
GROUP BY "frp"."id" 
HAVING "count" > $4 
ORDER BY "count" DESC, "name" ASC, "frp"."id" ASC
```

that's not valid. you cannot use projection (SELECT) columns in a having in PG, it must be:

```
GROUP BY "frp"."id"
   having COUNT(DISTINCT m.fileid) > 0
```

So Hacking the query with: 

```php
        if ($show_clusters && !$show_singles) {
            // doesnt work:
            //$query->having($query->expr()->gt($newCount, $query->createNamedParameter(1)));
            // maybe use $count from above? but that generates a alias AS "count" which me need to remove and 
            $query->having("count(DISTINCT m.fileid) > 0");
        }
```

Now it works on my machine :) 

Maybe that helps debugging it or solving the issues

**Comment by matiasdelellis** [comment]
Hi @zealot128 

I don't have much experience in Progress, from what little I know, it makes sense..  Don't you dare to make a PULL Requests with the change? 😉 

About:

> $query->having("count(DISTINCT m.fileid) > 0");

Must be:

> $query->having("count(DISTINCT m.fileid) > 1");

I prefer to use the alias, but obviously that is what fails.. 😅 

About:
```
-             // $query->groupBy('frp.name');  <---- doesnt work
+            $query->groupBy('frp.id');  // <---------------------
```
I'm not sure.. I understand that it would change how the query works..  Here I want you to group by name since 'Fri' can have several 'ids', with the same 'name'. 🤔

**Comment by zealot128** [comment]
@matiasdelellis 

no PR for now :) I've disabled faces for me (And postponed NC/memories at the same time), as it is not functional enough for me - tried around with the weights for the whole sunday, but either I get +7000 clusters, or ALL children plus asian Faces are in one group which is rather pointless for a faces search (tried weights between 0.35 and 0.45) - Back to iPhotos for now and repolishing my own (Rubyonrails/Vue.js) image gallery, which used to use Amazon Rekognition and a wizard that is able to select/deselect individual faces instead of relying on the clusters. Nevermind, i really dig your Vue.js Timeline scroller etc. component, that works great! I will try to regularly check the great work happening in the NC photos community.

About the grouping:

If you want to group by "name" AND select "frp.id" at the same time, what should the value be for "frp.id" if there could be multiple "frp.id" with the same names? That situation is undecidable, so PG throws an error (Also MySQL ...).

Normally, you can only select columns that either are:
- Grouped by 
- use an aggregation (e.g. ``array_agg(frp.id)`` to get an array of all frp.ids for that grouped name)
- OR if you group by "id" (primary key) you can also select columns that are in the same table because it is unambiguous  

So either: group by id (="person"), or by name but than you cannot select the frp.id.

**Comment by matiasdelellis** [comment]
@zealot128 
Thank you very much for the explanations. You've already been a big help. 
Enabling strict mode, I was able to reproduce the error. I hope to fix it soon. 😉 

p.s.: Do not apologize for deactivating it, we are aware that we cannot compete with google or amazon..  🙈  😉

**Comment by zealot128** [comment]
@matiasdelellis haha Amazon... Amazon Prime's Photos is not great either, puts all children in the same group, too, without any good Face management tool.  Also useless for me :) This is a strong suite of Apple, they will prompt you to help classify from time to time, which is the only way, especially with siblings.
Also regarding Amazon Photos, the date extraction for videos was bad, too, a lot of movies uploaded where put into "today", even when they had EXIF/XMP tags...

**Comment by harrysummer** [comment]
Great thanks for the discussion. After reading it, I made some change to the code and made it working with my local installation with PostgreSQL. The root causes of the error are:

1. The query selects non-group-by fields which causes confusion to PostgreSQL because the field can potentially have multiple different values for each group. I think MySQL would randomly return any value of the fields. We can surround the fields with an aggregate function to work around that.
2. PostgreSQL refuses to use the field alias in HAVING clause, so it needs to be expanded, as also mentioned in this thread above.

I didn't test on MySQL though.

```diff
--- TimelineQueryPeopleFaceRecognition_old.php  2023-02-01 19:23:40.106007245 +0000
+++ TimelineQueryPeopleFaceRecognition.php      2023-02-01 19:23:19.658062147 +0000
@@ -65,7 +65,7 @@

         // SELECT all face clusters
         $count = $query->func()->count($query->createFunction('DISTINCT m.fileid'), 'count');
-        $query->select('frp.id', 'frp.user as user_id', 'frp.name', $count)->from('facerecog_persons', 'frp');
+        $query->select($query->createFunction('min(`frp`.`id`) as `id`'), $query->createFunction('min(`frp`.`user`) as `user_id`'), 'frp.name', $count)->from('facerecog_persons', 'frp');

         // WHERE there are faces with this cluster
         $query->innerJoin('frp', 'facerecog_faces', 'frf', $query->expr()->eq('frp.id', 'frf.person'));
@@ -94,7 +94,7 @@

         // By default hides individual faces when they have no name.
         if ($show_clusters && !$show_singles) {
-            $query->having($query->expr()->gt('count', $query->createNamedParameter(1)));
+            $query->having($query->expr()->gt($query->func()->count($query->createFunction('DISTINCT m.fileid')), $query->createNamedParameter(1)));
         }

         // By default it shows the people who were not hidden
@@ -105,7 +105,7 @@
         // ORDER by number of faces in cluster
         $query->orderBy('count', 'DESC');
         $query->addOrderBy('name', 'ASC');
-        $query->addOrderBy('frp.id'); // tie-breaker
+        $query->addOrderBy('id'); // tie-breaker

         // FETCH all faces
         $cursor = $this->executeQueryWithCTEs($query);
```

**Comment by matiasdelellis** [comment]
Hello everyone,
Please can you test my PR at https://github.com/pulsejet/memories/pull/391 ? 😉 

I still couldn't install postgresql, and it was developed with strict mode in mariadb:... 🙈

**Comment by steffenmalisi** [comment]
@matiasdelellis Unfortunately I can **not** confirm that the changes from the PR resolve the issue. The initial error is gone, but another one is showing up. Using v4.11.2 of Memories with your changes included.

```
[index] Error: OC\DB\Exceptions\DbalException: An exception occurred while executing a query: SQLSTATE[42804]: Datatype mismatch: 7 ERROR:  argument of AND must be type boolean, not type bigint
LINE 28: ..." = $3) GROUP BY "frp"."id", "frp"."user" HAVING (COUNT(DIST...
                                                              ^ at <<closure>>

 0. /var/www/html/lib/private/DB/ConnectionAdapter.php line 69
    OC\DB\Exceptions\DbalException::wrap(["Doctrine\\DBAL ... "])
 1. /var/www/html/custom_apps/memories/lib/Db/TimelineQueryDays.php line 328
    OC\DB\ConnectionAdapter->executeQuery("WITH RECURSIVE  ... 5", [4,[3569],false,1,true], [2,101,5,2,2])
 2. /var/www/html/custom_apps/memories/lib/Db/TimelineQueryPeopleFaceRecognition.php line 213
    OCA\Memories\Db\TimelineQuery->executeQueryWithCTEs(["OC\\DB\\QueryBuilder\\QueryBuilder"])
 3. /var/www/html/custom_apps/memories/lib/Db/TimelineQueryPeopleFaceRecognition.php line 65
    OCA\Memories\Db\TimelineQuery->getFaceRecognitionClusters(["OCA\\Memories\\Db\\TimelineRoot"], 4, false, false)
 4. /var/www/html/custom_apps/memories/lib/Controller/PeopleController.php line 140
    OCA\Memories\Db\TimelineQuery->getPeopleFaceRecognition(["OCA\\Memories\\Db\\TimelineRoot"], 4, true)
 5. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 225
    OCA\Memories\Controller\PeopleController->facerecognitionPeople()
 6. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 133
    OC\AppFramework\Http\Dispatcher->executeController(["OCA\\Memories\ ... "], "facerecognitionPeople")
 7. /var/www/html/lib/private/AppFramework/App.php line 172
    OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Memories\ ... "], "facerecognitionPeople")
 8. /var/www/html/lib/private/Route/Router.php line 298
    OC\AppFramework\App::main("OCA\\Memories\\ ... r", "facerecognitionPeople", ["OC\\AppFramewo ... "], ["memories.People.facerecognitionPeople"])
 9. /var/www/html/lib/base.php line 1047
    OC\Route\Router->match("/apps/memories/ ... e")
10. /var/www/html/index.php line 36
    OC::handleRequest()

GET /index.php/apps/memories/api/facerecognition/people
```

**Comment by matiasdelellis** [comment]
Hi @steffenmalisi 
Please, try applying #417

**Comment by steffenmalisi** [comment]
Hey @matiasdelellis, wow thanks for the ultra fast response :-). With your suggested changes, People are now showing up also in the memories app!
